File: backend/northstar_mcp_typescript/github_ops.ts
/** GitHub operations for creating pull requests. */

/**
 * Create a pull request on GitHub, or return existing PR URL if one exists.
 *
 * @param repoFullname Repository in format 'owner/repo'
 * @param headBranch Source branch (the new changes)
 * @param baseBranch Target branch (usually 'main')
 * @param title PR title
 * @param body PR description
 * @returns URL of the created or existing pull request
 * @throws Error if PR creation fails
 */
export async function openPR(
  repoFullname: string,
  headBranch: string,
  baseBranch: string,
  title: string,
  body: string
): Promise<string> {
  const githubToken = Deno.env.get("GITHUB_TOKEN");

  if (!githubToken) {
    throw new Error(
      "GITHUB_TOKEN not found in environment.\n" +
        "Hint: Set it in your .env file or environment variables."
    );
  }

  try {
    const [owner, repo] = repoFullname.split("/");
    if (!owner || !repo) {
      throw new Error(
        `Invalid repository format: ${repoFullname}. Expected 'owner/repo'`
      );
    }

    // Check if PR already exists for this branch
    const existingPRsUrl = `https://api.github.com/repos/${repoFullname}/pulls?state=open&head=${owner}:${headBranch}`;
    const existingPRsResponse = await fetch(existingPRsUrl, {
      headers: {
        Authorization: `token ${githubToken}`,
        Accept: "application/vnd.github.v3+json",
      },
    });

    if (existingPRsResponse.ok) {
      const existingPRs = await existingPRsResponse.json();
      if (Array.isArray(existingPRs) && existingPRs.length > 0) {
        return existingPRs[0].html_url;
      }
    }

    // Create new PR
    const createPRUrl = `https://api.github.com/repos/${repoFullname}/pulls`;
    const createPRResponse = await fetch(createPRUrl, {
      method: "POST",
      headers: {
        Authorization: `token ${githubToken}`,
        Accept: "application/vnd.github.v3+json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title,
        body,
        base: baseBranch,
        head: headBranch,
      }),
    });

    if (!createPRResponse.ok) {
      const errorData = await createPRResponse.json().catch(() => ({}));
      const status = createPRResponse.status;
      let errorMsg = `GitHub API error: ${errorData.message || createPRResponse.statusText}`;

      if (status === 401) {
        errorMsg += "\nHint: Your GITHUB_TOKEN may be invalid or expired.";
      } else if (status === 404) {
        errorMsg +=
          `\nHint: Repository '${repoFullname}' not found or you don't have access.`;
      } else if (status === 422) {
        errorMsg +=
          "\nHint: Check that base and head branches are different and exist.";
      }

      throw new Error(errorMsg);
    }

    const prData = await createPRResponse.json();
    return prData.html_url;
  } catch (e) {
    if (e instanceof Error) {
      throw e;
    }
    throw new Error(`Failed to create pull request: ${String(e)}`);
  }
}

File: backend/northstar_mcp/github_ops.py
"""GitHub operations for creating pull requests."""

import os
from github import Github, GithubException


def open_pr(
    repo_fullname: str,
    head_branch: str,
    base_branch: str,
    title: str,
    body: str
) -> str:
    """
    Create a pull request on GitHub, or return existing PR URL if one exists.

    Args:
        repo_fullname: Repository in format 'owner/repo'
        head_branch: Source branch (the new changes)
        base_branch: Target branch (usually 'main')
        title: PR title
        body: PR description

    Returns:
        URL of the created or existing pull request

    Raises:
        Exception: If PR creation fails
    """
    github_token = os.getenv("GITHUB_TOKEN")

    if not github_token:
        raise Exception(
            "GITHUB_TOKEN not found in environment.\n"
            "Hint: Set it in your .env file or environment variables."
        )

    try:
        g = Github(github_token)
        repo = g.get_repo(repo_fullname)

        # Check if PR already exists for this branch
        existing_prs = repo.get_pulls(state="open", head=f"{repo.owner.login}:{head_branch}")

        for pr in existing_prs:
            return pr.html_url

        # Create new PR
        pr = repo.create_pull(
            title=title,
            body=body,
            base=base_branch,
            head=head_branch
        )

        return pr.html_url

    except GithubException as e:
        error_msg = f"GitHub API error: {e.data.get('message', str(e))}"

        if e.status == 401:
            error_msg += "\nHint: Your GITHUB_TOKEN may be invalid or expired."
        elif e.status == 404:
            error_msg += f"\nHint: Repository '{repo_fullname}' not found or you don't have access."
        elif e.status == 422:
            error_msg += "\nHint: Check that base and head branches are different and exist."

        raise Exception(error_msg)
    except Exception as e:
        raise Exception(f"Failed to create pull request: {str(e)}")
